tom/changelog.py

```diff
@@ -0,0 +1,159 @@
+            if m:
+                return ('\n'.join(changelog_lines[:i]),
+                        m.group(1),
+                        '\n'.join(changelog_lines[i:]))
+        raise ChangelogParsingException('Failed to find version in changelog file, top 5 lines: {}'.format(changelog_lines[:5]))
+
+    def get_next_version(self, old_version, branch):
+        version_parts = old_version.split('.')
+        if branch == 'master':
+            try:
+                # increase minor version and set patch version to 0,
+                # i.e. 3.10.2 -> 3.11.0
+                version_parts[1] = str(int(version_parts[1]) + 1)
+                version_parts[2] = '0'
+            except:
+                # ???
+                raise
+        else:
+            if branch == '{}.{}.x'.format(old_version_parts[0], old_version_parts[1]):
+                # we already had version corresponding to a branch name,
+                # i.e. version 3.10.2 on a branch 3.10.x
+                # so increase patch version
+                # i.e. 3.10.2 -> 3.10.3
+                version_parts[2] = str(int(version_parts[2]) + 1)
+            else:
+                # this is a first version on this branch
+                # so set it to [branch_name].0
+                branch_name_parts = branch.split('.')
+                version_parts = branch_name_parts[0:2] + [0]
+        return '.'.join(version_parts)
+
+    def get_changelog_for(self, name, arg, old_version, branch):
+        olddir = os.getcwd()
+        try:
+            os.chdir(os.path.join(self.repos_root, name))
+            cmd = ['../core/misc/changelog-generator/changelog-generator',
+                    arg,
+                    '{}...{}'.format(old_version, branch)]
+            log.debug('running command: {}'.format(' '.join(cmd)))
+            proc = subprocess.run(cmd,
+                    stdout=subprocess.PIPE,
+                    universal_newlines=True,
+                    check=True)
+        finally:
+            os.chdir(olddir)
+        return proc.stdout
+
+    def generate_changelog_in_repo(self, branch, repo):
+        name = repo.repo_name
+        changelog_filename = self.changelog_filenames.get(name)
+        if not changelog_filename:
+            # no need to generate changelog in _this_ repo
+            return False
+
+        # find old and new versions
+        (header, old_version, old_changelog) = self.split_changelog_into_parts(changelog_filename, repo)
+        if header != '':
+            header += '\n'
+        new_version = self.get_next_version(old_version, branch)
+        log.debug('next version: {} => {}'.format(old_version, new_version))
+
+        # checkout new branch
+        timestamp = re.sub('[^0-9-]', '_', str(datetime.datetime.today()))
+        new_branchname = '{}-changelog-{}'.format(new_version, timestamp)
+        repo.checkout(new_branchname, True)
+
+        if name != 'enterprise':
```

Comment:
If this whole class is supposed to be cfengine specific, maybe add a config option for it? or at least a note?

Reply:
Good point. I was following this guide when writing this function: https://github.com/mendersoftware/infra/blob/master/files/buildcache/release-scripts/RELEASE_PROCESS.org#generate-changelogs

Do you have something like this for Mender?